src/WebJobs.Script.Extensibility/ScriptBindingProvider.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Reflection;
+using Microsoft.Azure.WebJobs.Host;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.WebJobs.Script.Extensibility
+{
+    /// <summary>
+    /// Base class for providers of <see cref="ScriptBinding"/>s.
+    /// </summary>
+    public abstract class ScriptBindingProvider
+    {
+        /// <summary>
+        /// Constructs a new instance.
+        /// </summary>
+        /// <param name="config">The <see cref="JobHostConfiguration"/>.</param>
+        /// <param name="hostMetadata">The host configuration metadata.</param>
+        /// <param name="traceWriter">The <see cref="TraceWriter"/> that can be used to log trace events.</param>
+        public ScriptBindingProvider(JobHostConfiguration config, JObject hostMetadata, TraceWriter traceWriter)
+        {
+            Config = config;
+            Metadata = hostMetadata;
+            TraceWriter = traceWriter;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="JobHostConfiguration"/>.
+        /// </summary>
+        protected JobHostConfiguration Config { get; private set; }
+
+        /// <summary>
+        /// Gets the host configuration metadata.
+        /// </summary>
+        protected JObject Metadata { get; private set; }
+
+        /// <summary>
+        /// Gets the <see cref="TraceWriter"/> that can be used to log trace events.
+        /// </summary>
+        protected TraceWriter TraceWriter { get; private set; }
+
+        /// <summary>
+        /// Called after all bindings have been created to allow the provider
+        /// to perform final host level initialization, extension registration, etc.
+        /// </summary>
+        public virtual void Initialize()
+        {
+        }
+
+        /// <summary>
+        /// Create a <see cref="ScriptBinding"/> for the specified metadata if
+        /// </summary>
+        /// <param name="bindingMetadata">The metadata for the binding.</param>
+        /// <param name="binding">The corresponding <see cref="ScriptBinding"/>.</param>
+        /// <returns>True if a binding was created, false otherwise.</returns>
+        public abstract bool TryCreate(JObject bindingMetadata, out ScriptBinding binding);
```

Comment:
It would be nice to pass a type that contains the raw metadata, but also exposes properties for the known things like type, direction, data type, etc.